src/compiler/scala/tools/nsc/transform/Fields.scala

```diff
@@ -588,75 +589,81 @@ abstract class Fields extends InfoTransform with ast.TreeDSL with TypingTransfor
     }
 
     // synth trees for accessors/fields and trait setters when they are mixed into a class
-    def fieldsAndAccessors(clazz: Symbol): List[ValOrDefDef] = {
-      def fieldAccess(accessor: Symbol): List[Tree] = {
-        val fieldName = accessor.localName
-        val field = clazz.info.decl(fieldName)
-        // The `None` result denotes an error, but it's refchecks' job to report it (this fallback is for robustness).
-        // This is the result of overriding a val with a def, so that no field is found in the subclass.
-        if (field.exists) List(Select(This(clazz), field))
-        else Nil
-      }
-
-      def getterBody(getter: Symbol): List[Tree] = {
+    def fieldsAndAccessors(clazz: Symbol): List[Tree] = {
+      // scala/scala-dev#219
+      // Cast to avoid spurious mismatch in paths containing trait vals that have
+      // not been rebound to accessors in the subclass we're in now.
+      // For example, for a lazy val mixed into a class, the lazy var's info
+      // will not refer to symbols created during our info transformer,
+      // so if its type depends on a val that is now implemented after the info transformer,
+      // we'll get a mismatch when assigning `rhs` to `lazyVarOf(getter)`.
+      // TODO: could we rebind more aggressively? consider overriding in type equality?
+      def cast(tree: Tree, pt: Type) = gen.mkAsInstanceOf(tree, pt)
```

Comment:
Since this type checked without the cast for the whole community build
until we hit the glitch in the test case, the cast should get dropped by
erasure (or genbcode?). The only reason it doesn't type check before
erasure is explained in the PR. I don't feel strongly about using the
cast-based approach or not. It's probably slightly more robust than the
previous commit, but you're right we should check the byte code. I did a
spot check but not the usual full diff of a bootstrap.

On Wed, Sep 14, 2016 at 17:24 Lukas Rytz <notifications@github.com> wrote:

> In src/compiler/scala/tools/nsc/transform/Fields.scala
> <https://github.com/scala/scala/pull/5388#discussion_r78771687>:
>
> > -        // This is the result of overriding a val with a def, so that no field is found in the subclass.
> > -        if (field.exists) List(Select(This(clazz), field))
> > -        else Nil
> > -      }
> > -
> > -      def getterBody(getter: Symbol): List[Tree] = {
> > +    def fieldsAndAccessors(clazz: Symbol): List[Tree] = {
> > +      // scala/scala-dev#219
> > +      // Cast to avoid spurious mismatch in paths containing trait vals that have
> > +      // not been rebound to accessors in the subclass we're in now.
> > +      // For example, for a lazy val mixed into a class, the lazy var's info
> > +      // will not refer to symbols created during our info transformer,
> > +      // so if its type depends on a val that is now implemented after the info transformer,
> > +      // we'll get a mismatch when assigning `rhs` to `lazyVarOf(getter)`.
> > +      // TODO: could we rebind more aggressively? consider overriding in type equality?
> > +      def cast(tree: Tree, pt: Type) = gen.mkAsInstanceOf(tree, pt)
>
> this seems to introduce quite a number of casts, it looks like rather big
> hammer. maybe there's a potential that it hides latent issues (now or in
> the future). we should also check if these casts are eliminated in erasure,
> or if some remain..
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/5388/files/4d67c390d989e8ecf3b3916465dea1b0b70fde57..8b79ea9415ac99ebabe29577cd83ec5081d2f488#r78771687>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFjy7pGI36xDeA6XkLcovHeRQrZzn1dks5qqBHGgaJpZM4J4MQp>
> .
>